### PR TITLE
AArch64: Changes to combine shift and negate to a single instruction.

### PR DESCRIPTION
**Pull request by AndyJGraham**
Add three node match rules to handle the three varieties of shift, together with capability to emit a combined negate instruction with shift.

### CODE REVIEW

**Comment by AndyJGraham** [comment]
> Hello Andrew Graham, thanks for contributing a PR to our project!
> 
> We use the [Oracle Contributor Agreement](http://www.oracle.com/technetwork/community/oca-486395.html) to make the copyright of contributions clear. We don't have a record of you having signed this yet, based on your email address andrew -(dot)- graham -(at)- arm -(dot)- com. You can sign it at that link.
> 
> If you think you've already signed it, please comment below and we'll check.

Hi team, I’m an employee of Arm and I’ll be contributing under their signature. Please let me know if there is anything else I need to do there.

File: compiler/src/org.graalvm.compiler.core.aarch64.test/src/org/graalvm/compiler/core/aarch64/test/AArch64BitFieldTest.java [hunk 1]
```diff
@@ -207,7 +207,7 @@ public void testSignedI2LInsert() {
     // SBFIZ with B2I.
     public int signedB2IInsert(int input) {
         byte b = (byte) input;
-        return b << 31;
+        return b << 30;
```
Comment (teshull) [inline compiler/src/org.graalvm.compiler.core.aarch64.test/src/org/graalvm/compiler/core/aarch64/test/AArch64BitFieldTest.java:210]: Out of curiosity, why did you change this?

File: compiler/src/org.graalvm.compiler.core.aarch64/src/org/graalvm/compiler/core/aarch64/AArch64NodeMatchRules.java [hunk 2]
```diff
@@ -521,6 +522,26 @@ public ComplexMatchResult logicShift(BinaryNode binary, ValueNode a, BinaryNode
...
+    @MatchRule("(Negate (RightShift=shift a Constant=b))")
+    @MatchRule("(Negate (LeftShift=shift a Constant=b))")
+    public ComplexMatchResult negShift(BinaryNode shift, ValueNode a, ConstantNode b) {
+        assert b.getStackKind().isNumericInteger();
+        assert a.getStackKind().isNumericInteger();
+        int maxAllowed = shift.getStackKind().getBitCount() - 1;
+        int shiftAmt = b.asJavaConstant().asInt();
+        if (shiftAmt > maxAllowed || shiftAmt < 0) {
+            return null;
+        }
```
Comment (teshull) [inline compiler/src/org.graalvm.compiler.core.aarch64/src/org/graalvm/compiler/core/aarch64/AArch64NodeMatchRules.java:535]: By clamping the value within the macroassembler, there shouldn't be the need for the above range check and bailout

**Review by teshull — CHANGES_REQUESTED** [review]

File: compiler/src/org.graalvm.compiler.asm.aarch64/src/org/graalvm/compiler/asm/aarch64/AArch64MacroAssembler.java [hunk 3]
```diff
@@ -966,6 +966,19 @@ public void neg(int size, Register dst, Register src) {
...
+     * dst = -(shiftType(src, shiftAmt & (size - 1))).
+     *
+     * @param size register size. Has to be 32 or 64.
+     * @param dst general purpose register. May not be null or stackpointer.
+     * @param src general purpose register. May not be null or stackpointer.
+     * @param shiftType right or left shift, arithmetic or logical.
+     * @param shiftAmt number of shift bits. Has to be between 0 and (size - 1).
+     */
+    public void neg(int size, Register dst, Register src, ShiftType shiftType, int shiftAmt) {
+        sub(size, dst, zr, src, shiftType, shiftAmt);
```
[the remark this answers is not in the document]
Comment (AndyJGraham) [reply compiler/src/org.graalvm.compiler.asm.aarch64/src/org/graalvm/compiler/asm/aarch64/AArch64MacroAssembler.java:979]: This change has now been implemented.

File: compiler/src/org.graalvm.compiler.core.aarch64/src/org/graalvm/compiler/core/aarch64/AArch64NodeMatchRules.java (hunk 2 above)
Comment (AndyJGraham) [reply compiler/src/org.graalvm.compiler.core.aarch64/src/org/graalvm/compiler/core/aarch64/AArch64NodeMatchRules.java:535]: This change has been implemented.

File: compiler/src/org.graalvm.compiler.core.aarch64.test/src/org/graalvm/compiler/core/aarch64/test/AArch64BitFieldTest.java (hunk 1 above)
Comment (AndyJGraham) [reply compiler/src/org.graalvm.compiler.core.aarch64.test/src/org/graalvm/compiler/core/aarch64/test/AArch64BitFieldTest.java:210]: Investigation took place to determine why and how a negate and shift was generated when there is no apparent negate in signedB2IInsert().

It's quite interesting how we get negate+shift when there is no apparent negate in signedB2IInsert(). The following optimizations result in negate+shift generated for the same operation.

Reassociation phase replaces the left shift (input << 31) with multiply (input * -2147483648). This simplification occurs from tryReplaceWithMulNode() in ReassociatePhase.java Line#163. As 2's complement of 2 << 31 is same as its negative in 4 byte representation, up-down between int and long results in selecting a -ve multiplier. 2. Cannonicalizer phase replaces the multiply with shift+negate in cannonical() from MulNode.java Line#161.

Therefore, as Xiaohong Gong suggested, we simply updated the actual test to use shift with a smaller number. Therefore, we now shift by << 30 instead of 31.

**Comment by teshull** [comment]
looks good to me. I'll put it into the gates now
